src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs

```diff
@@ -138,33 +141,46 @@ private void CheckInvariants(SyntaxToken token, VirtualCharSequence result)
 
             for (var index = startIndexInclusive; index < endIndexExclusive;)
             {
-                if (tokenText[index] == '"' &&
-                    tokenText[index + 1] == '"')
+                if (tokenText[index] == '"' && tokenText[index + 1] == '"')
                 {
                     result.Add(new VirtualChar('"', new TextSpan(offset + index, 2)));
                     index += 2;
                 }
-                else if (escapeBraces &&
-                            (tokenText[index] == '{' || tokenText[index] == '}'))
+                else if (escapeBraces && IsOpenOrCloseBrace(tokenText[index]))
                 {
-                    if (!TryAddBraceEscape(result, tokenText, offset, index))
-                    {
+                    if (!IsLegalBraceEscape(tokenText, offset, index, out var span))
                         return default;
-                    }
 
+                    result.Add(new VirtualChar(tokenText[index], span));
                     index += result.Last().Span.Length;
                 }
+                else if (Rune.TryCreate(tokenText[index], out var rune))
+                {
+                    // First, see if this was a single char that can become a rune (the common case).
+                    result.Add(new VirtualChar(rune, new TextSpan(offset + index, 1)));
+                    index += 1;
+                }
+                else if (index + 1 < tokenText.Length &&
+                         Rune.TryCreate(tokenText[index], tokenText[index + 1], out rune))
+                {
+                    // Otherwise, see if we have a surrogate pair (less common, but possible).
+                    result.Add(new VirtualChar(rune, new TextSpan(offset + index, 2)));
+                    index += 2;
+                }
                 else
                 {
-                    result.Add(new VirtualChar(tokenText[index], new TextSpan(offset + index, 1)));
-                    index++;
+                    // Something that couldn't be encoded as runes.
+                    return default;
```

Comment:
💭 I would prefer to still return a `VirtualCharSequence`, with the following characteristics:

* Unmatched surrogates are replaced with the fallback character, but still identify the correct span in the original source
* The `VirtualCharSequence` has a flag to indicate that the text contents of the string could not be fully-interpreted, and thus the sequence shouldn't be used for lossless transformations like converting a string literal to a verbatim string literal

Reply:
Great idea.  I'll do ths.

One niggle is that you can trivial (i.e. O(1) cost) create a subsequence of a virtualcharsequence.  It will not be cheap to do that if we have to have tha value be correct on subsequences as we'll have to compute if the subsequence contains the invalid char.  if we're ok with the subsequences just inheriting htat value from the main sequence, this is easy to do.